AI/VCAI/Pathfinding/AIPathfinder.cpp

```diff
@@ -84,13 +73,43 @@ std::shared_ptr<AINodeStorage> AIPathfinder::getOrCreateStorage(const HeroPtr &
 
 		auto config = std::make_shared<AIPathfinding::AIPathfinderConfig>(cb, ai, nodeStorage);
 
+		logAi->debug("Recalculate paths for %s", hero->name);
 		cb->calculatePaths(config, hero.get());
 	}
+}
+
+void AIPathfinder::updatePaths(const HeroPtr & hero)
+{
+	std::shared_ptr<AINodeStorage> nodeStorage;
+
+	if(!vstd::contains(storageMap, hero))
+	{
+		if(storageMap.size() < storagePool.size())
```

Comment:
This duplicates code in `AIPathfinder::updatePaths(std::vector<HeroPtr>)`.

Reply:
I know. What do you think about parallel calculation of paths for all heroes? Can you suggest how it can be implemented? I mean is there any thread pooling tool or some libraries?

Reply:
Can I try this https://en.cppreference.com/w/cpp/thread/async

Reply:
There is `CThreadHelper` in vcmi, it is very simple but should be sufficient for such task.

Reply:
Also there is a thread pool in boost, you should check it out.

Reply:
How can I determine the amount of threads for CThreadHelper? I expect it should equal to amount of CPU cores?

Reply:
Or maybe amount of heroes. It should not be more than 8. Or minimum between the first and the second.

Reply:
Use `boost::thread::hardware_concurrency()` for amount of threads. There is a complete usage example in `CBattleAI::attemptCastingSpell()`

Reply:
_minimum between the first and the second_ sounds reasonable

Reply:
Can logger be accessed concurrently?

Reply:
Yes it can.